org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/AST.java

```diff
@@ -3257,8 +3257,8 @@ public TypeDeclaration newTypeDeclaration() {
 	public TypeDeclarationStatement
 			newTypeDeclarationStatement(AbstractTypeDeclaration decl) {
 		TypeDeclarationStatement result = new TypeDeclarationStatement(this);
-		if (this.apiLevel == AST.JLS2) {
-			result.internalSetTypeDeclaration((TypeDeclaration) decl);
+		if (this.apiLevel == AST.JLS2 && decl instanceof TypeDeclaration namedType) {
```

Comment:
Would you please open a PR against jdt.core master for that part? It would be best to do such changes upstream too to ease merging/rebasing.